doc/man/pc_query.1
'\" t
.\"     Title: pc_query
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 01/31/2014
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PC_QUERY" "1" "01/31/2014" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pc_query \- query the local address book
.SH "SYNOPSIS"
.sp
\fBpc_query\fR [\fIOPTIONS\fR] [SEARCHSTRING]
.SH "DESCRIPTION"
.sp
pc_query(1) prints contacts from a local address book, synchronized with remote CardDAV resources with pycardsyncer(1)\&. Contact information is kept in the \fIvCard\fR format, one set of information, corresponding to a business card, is therefore often referred to as a \fIcard\fR\&. pc_query(1) can also import and export vcards from/to the database\&. pc_query(1) is part of the \fIpycarddav\fR package\&.
.SH "OPTIONS"
.SS "BASIC STARTUP OPTIONS"
.PP
\fB\-a, \-\-account\fR=\fINAME\fR
.RS 4
Only use the account/resource
\fINAME\fR
(this option can be used more than once)\&.
.RE
.PP
\fB\-c, \-\-config\fR=\fICONFIG\fR
.RS 4
Use the configuration file
\fICONFIG\fR, otherwise pc_query(1) will look in
\fI$HOME/\&.pycard/\fR
and
\fI$HOME/\&.config/pycard/\fR
for files named
\fIpycard\&.conf\fR\&.
.RE
.PP
\fB\-\-debug\fR
.RS 4
This option enables debugging output\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Print a small help text and exit\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Print pc_query\(cqs version number and exit\&.
.RE
.SS "PRINTING OPTIONS"
.sp
These options can only be used with \fISEARCHSTRING\fR and determine the output format of pc_query(1)\&. Without any options pc_query(1) will print the name, telephone numbers and email addresses of all matching contact cards\&.
.PP
\fB\-A\fR
.RS 4
Prints the whole card\&.
.RE
.PP
\fB\-m\fR
.RS 4
Only prints names and email addresses, in a mutt friendly format (one line per email address)\&.
.RE
.PP
\fB\-t\fR
.RS 4
Only prints names and telephone numbers, analogue to \-m (but in different sequence)
.RE
.SS "FILE INPUT AND OUTPUT OPTIONS"
.PP
\fB\-b, \-\-backup\fR\fI=\*(AqFILE\fR
.RS 4
Backup the local db to BACKUP, if a SEARCHSTRING is present, only backup cards matching it\&.
.RE
.PP
\fB\-\-delete\fR
.RS 4
Delete card matching
\fISEARCHSTRING\fR, if more than one matches, the user has to choose one card in an interactive user interface\&.
.RE
.PP
\fB\-i, \-\-import\fR=\fIFILE\fR
.RS 4
Import vCard from FILE or
\fISTDIN\fR
into the first specified (or default) account
.RE
.SS "OTHER OPTIONS"
.PP
\fB\-\-where\fR=\fIMODE\fR
.RS 4
Decides which part of the contact cards are matched against the searchterm\&.
\fIMODE\fR
is one of
\fIvcard\fR,
\fIname\fR
(the structured name, as used in the vcard specification),
\fIfname\fR
(the formated name, meaning as is printed by pc_query) and
\fIallnames\fR
(\fIallnames\fR
means search
\fIfname\fR
as well as
\fIname\fR)\&. While only search through the names might not find what you are looking for, it will considerably speed up your query\&.
.RE
.SH "AUTHOR"
.sp
pc_query was mostly written by Christian Geier, with a lot of help by others, see \fICONTRIBUTORS\&.txt\fR in the pycarddav distribution\&.
.SH "RESOURCES"
.sp
Main web site: http://lostpackets\&.de/pycarddav/ Please report bugs via the contact information at the above web site or via github: http://github\&.com/geier/khal/\&.
.SH "SEE ALSO"
.sp
pycardsyncer(1), pycard\-import(1)
.SH "COPYING"
.sp
Copyright (C) 2011\-2013 Christian Geier and Contributors\&. pc_query and pycarddav are released under the terms of the Expat/MIT license, see the \fICOPYING\fR file distributed with pycarddav\&.


NEWS
News
====

15.06.2013: pyCardDAV v.0.5.0 released

 **NEW** This release brings support for multiple CardDAV accounts. See the
 usage instructions, the supplied example config and/or pc_query --help. Also
 support for more CardDAV servers is included. If you are upgrading, you need to
 delete the local database, otherwise pyCardDAV will refuse to work.

28.03.2013: pyCardDAV v0.4.1 released

  pyCardDAV v0.4.1 is released. This is a bugfix release, if everything works
  fine for you, there is no need to upgrade. Database deletion should not be
  necessary.

15.11.2012: pyCardDAV v0.4 released

  pyCardDAV v0.4 is released. This is a mayor rewrite (again), so some previously
  fixed bugs might be back in. If your upgrading, you should delete your database
  file first.

  On the plus side, there are some new features in pyCardDAV:

    experimental write support in the backend
    import & export vCards
    import addresses directly from mutt
    speed increase in (initial) sync due to switching from pycurl to requests
    detects removed cards on server and delete them locally
    can delete cards locally and then on server

  Also the license has changed to MIT/Expat, see the COPYING file for details
  (but a beer is still appreciated).

  PyCurl is not required anymore, pyCardDAV relies on requests now (which needs
  to be installed).

  Special thanks to David Soulayrol who made a lot of this happen.

  Attention: please make sure you have a backup when you enable write support, see
  Usage for more details.

27.01.2012: pyCardDAV v0.3.3_ released:
  **New** sabredav/owncloud support, thanks Davide Gerhard.

  Fixes a bug where properties with no type parameters were not printed.

  This release also fixes a small database bug.

  The config file has a new entry (*davserver*) which you can set to
  either davical or sabredav (depending on your CardDAV server).

  Future:
  The source code has been cleaned up quite a bit (nearly every line of code
  has been touched) and some features have been added. Write support is nearly
  finished in the backend (but will probably not be included in the next
  release yet), but the frontend is still really buggy and a pain to use. If
  you want to have a look, check the repository out at github and check the
  branch *write_support* (but it might me broken).

06.01.2012: pyCardDAV v0.3.2_ released:
  this is a minor bugfix update, db deleting should not be necessary. If
  everything is working fine at the moment, there is no need to upgrade.

06.01.2012: pyCardDAV v0.3.1_ released:
  this bugfix release fixes some bugs on Debian and a formatting bug (thanks to
  Antoine Sirinelli) and one more unicode bug (thanks to Thomas Klausner). Also,
  some more meaningful error messages were added.

  **Attention** if you are upgrading:
  you should delete the old database again and resync using pycardsyncer

08.12.2011: pyCardDAV v0.3_ released:
  this fixes an unicode bug and has a lot of internal changes

  **Attention** pc-query has been renamed to pc_query,
  make sure to delete the old database, also the config file
  format has somewhat changed

10.10.2011: pyCardDAV v0.2.1_ released
  this fixes a minor bug in the example config file

14.09.2011: pyCardDAV v0.2_ released
  **New** config files are now supported

13.09.2011: pyCardDAV moved to github_
  feel free to fork etc.

12.08.2011: pyCardDAV v0.1_ released
  first public version


pycarddav/pc_query.py
#!/usr/bin/env python
# vim: set ts=4 sw=4 expandtab sts=4:
# ----------------------------------------------------------------------------
# "THE BEER-WARE LICENSE" (Revision 42):
# <geier@lostpackets.de> wrote this file. As long as you retain this notice you
# can do whatever you want with this stuff. If we meet some day, and you think
# this stuff is worth it, you can buy me a beer in return Christian Geier
# ----------------------------------------------------------------------------

"""
utility for querying the database
"""
from __future__ import print_function

try:
    from os import path
    from pycarddav import Configuration, ConfigurationParser
    from pycarddav import capture_user_interruption
    from pycarddav import backend
    from pycarddav import model
    from pycarddav import ui

    import argparse
    import sys

except ImportError as error:
    sys.stderr.write(str(error))
    sys.exit(1)


class QueryConfigurationParser(ConfigurationParser):
    """A specialized setup tool for cards query."""
    def __init__(self, desc):
        ConfigurationParser.__init__(self, desc)

        self.set_mandatory_options([(Configuration.SECTIONS.DB, 'path')])

        self._arg_parser.add_argument(
            "-a", action="store_true", dest="cmd__display_all", default=False,
            help="prints the whole card, not only name, "
            "telephone numbers and email addresses")
        self._arg_parser.add_argument(
            "-m", action="store_true", dest="cmd__mutt", default=False,
            help="only prints email addresses, in a mutt friendly format")
        self._arg_parser.add_argument(
            "-e", action="store_true", dest="cmd__edit", default=False,
            help="edit the contact file.\n"
            "NOTE: this feature is experimental and will probably corrupt "
            "your *local* database. Your remote CardDAV resource will stay "
            "untouched, as long as You don't enable write support for the "
            "syncer.")
        self._arg_parser.add_argument(
            "cmd__search_string", metavar="SEARCHSTRING", default="",
            help="the string to search for", nargs="?")
        self._arg_parser.add_argument(
            "-b", "--backup", action="store", dest="cmd__backup",
            metavar="BACKUP",  help="backup the local db to BACKUP, "
            "if a SEARCHSTRING is present, only backup cards matching it.")
        self._arg_parser.add_argument(
            "-i", "--import", metavar="FILE",
            type=argparse.FileType("r"), dest="cmd__importing",
            help="import vcard from FILE or STDIN")
        self._arg_parser.add_argument(
            "--delete", dest="cmd__delete", action="store_true",
            help="delete card matching SEARCHSTRING")


def main():
    """main function, everything starts  here"""
    capture_user_interruption()

    # Read configuration.
    parser = QueryConfigurationParser('prints contacts cards matching a search string')

    conf = parser.parse()
    if conf is None:
        sys.exit(1)

    # testing if the db exists
    if not path.exists(conf.sqlite__path):
        sys.exit(str(conf.sqlite__path) + " file does not exist, please sync with "
                "pycardsyncer first.")

    search_string = conf.cmd__search_string.decode("utf-8")

    my_dbtool = backend.SQLiteDb(conf.sqlite__path, "utf-8", "stricts", False)

    #import:
    if conf.cmd__importing:
        cards = model.cards_from_file(conf.cmd__importing)
        for card in cards:
            my_dbtool.update(card, status=backend.NEW)
        sys.exit()

    # backup:
    if conf.cmd__backup:
        with open(conf.cmd__backup, 'w') as vcf_file:
            if search_string == "":
                hreflist = my_dbtool.get_all_vref_from_db()
            else:
                hreflist = my_dbtool.search(search_string)
            for href in hreflist:
                vcard = my_dbtool.get_vcard_from_db(href)
                vcf_file.write(vcard.vcf.encode('utf-8'))
        sys.exit()

    # editing a card:
    if conf.cmd__edit:
        names = my_dbtool.select_entry2(search_string)
        href = ui.select_entry(names)
        if href is None:
            sys.exit("Found no matching cards.")

    # print card(s)
    if conf.cmd__delete:
        names = my_dbtool.select_entry2(search_string)
        href = ui.select_entry(names)
        if href is None:
            sys.exit('Found no matching cards.')
        my_dbtool.mark_delete(href)
        print('vcard %s deleted from local db, will be deleted on ' % href + \
            'the server on the next sync')
        sys.exit()

    print("searching for " + conf.cmd__search_string + "...")
    result = my_dbtool.search(search_string)
    for one in result:
        vcard = my_dbtool.get_vcard_from_db(one)
        if conf.cmd__mutt:
            lines = vcard.print_email()
        elif conf.cmd__display_all:
            lines = vcard.pretty
        else:
            lines = vcard.pretty_min
        if not lines == '':
            print(lines.encode('utf-8'))

    return 0

if __name__ == "__main__":
    main()
